hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java

```diff
@@ -99,7 +99,177 @@ public SCMContainerPlacementRackAware(final NodeManager nodeManager,
+      mutableExcludedNodes.add(secondNode);
+    } else {
+
+      // choose node to meet replication requirement
+      // case 1: one used node, choose one on the same rack as the used
+      // node, choose others on different racks.
+      if (mutableUsedNodes.size() == 1) {
+        favoredNode = favoredNodeNum > favorIndex ?
+            mutableFavoredNodes.get(favorIndex) : null;
+        DatanodeDetails firstNode;
+        if (favoredNode != null &&
+            networkTopology.isSameParent(mutableUsedNodes.get(0),
+            favoredNode)) {
+          firstNode = favoredNode;
+          favorIndex++;
+        } else {
+          firstNode = chooseNode(mutableExcludedNodes, mutableUsedNodes,
+              metadataSizeRequired, dataSizeRequired);
+        }
+        chosenNodes.add(firstNode);
+        nodesRequired--;
+        if (nodesRequired == 0) {
+          return Arrays.asList(chosenNodes.toArray(new DatanodeDetails[0]));
+        }
+
+        // choose remaining nodes on different racks
+        mutableExcludedNodes.add(firstNode);
+        mutableExcludedNodes.addAll(mutableUsedNodes);
+      } else {
+        // case 2: two or more used nodes, if these two nodes are
+        // in the same rack, then choose nodes on different racks, otherwise,
+        // choose one on the same rack as one of used
+        // nodes, remaining chosen
+        // are on different racks.
+        for (int i = 0; i < usedNodesCount; i++) {
+          for (int j = i + 1; j < usedNodesCount; j++) {
+            if (networkTopology.isSameParent(
+                usedNodes.get(i), usedNodes.get(j))) {
+              // choose remaining nodes on different racks
+              mutableExcludedNodes.addAll(mutableUsedNodes);
+              return chooseNodes(mutableExcludedNodes, chosenNodes,
+                  mutableFavoredNodes, favorIndex, nodesRequired,
+                  metadataSizeRequired, dataSizeRequired);
+            }
+          }
+        }
+        // choose one data on the same rack with one used node
+        favoredNode = favoredNodeNum > favorIndex ?
+            mutableFavoredNodes.get(favorIndex) : null;
+        DatanodeDetails secondNode;
+        if (favoredNode != null && networkTopology.isSameParent(
+            chosenNodes.get(0), favoredNode)) {
+          secondNode = favoredNode;
+          favorIndex++;
+        } else {
+          secondNode =
+              chooseNode(mutableExcludedNodes, mutableUsedNodes,
+                  metadataSizeRequired, dataSizeRequired);
+        }
+        chosenNodes.add(secondNode);
+        mutableExcludedNodes.add(secondNode);
+        mutableExcludedNodes.addAll(mutableUsedNodes);
+        nodesRequired--;
+        if (nodesRequired == 0) {
+          return Arrays.asList(chosenNodes.toArray(new DatanodeDetails[0]));
+        }
+      }
+    }
+    // choose remaining nodes on different racks
+    return chooseNodes(mutableExcludedNodes, chosenNodes, mutableFavoredNodes,
```

Comment:
I am not sure if this is going to do what we want. The first few lines of that method are:

```
  private List<DatanodeDetails> chooseNodes(List<DatanodeDetails> excludedNodes,
      List<DatanodeDetails> chosenNodes, List<DatanodeDetails> favoredNodes,
      int favorIndex, int nodesRequired, long metadataSizeRequired,
      long dataSizeRequired) throws SCMException {
    Preconditions.checkArgument(chosenNodes != null);
    List<DatanodeDetails> excludedNodeList = excludedNodes != null ?
        excludedNodes : chosenNodes;
```
If we have excluded nodes and used nodes, then it sets the list to just excludedNodes - this is fine as you have added the chosen nodes to excluded on the way through this method.

Then it passes just the excludedNodes to:

```
chosenNode = chooseNode(excludedNodeList, null,
            metadataSizeRequired, dataSizeRequired);
```
Which eventually calls:

```
        node = (DatanodeDetails)networkTopology.chooseRandom(NetConstants.ROOT,
            excludedNodesForCapacity, excludedNodes, null, ancestorGen);
```

This is what I am not sure about - does that call above attempt to ensure that the picked node does not share a rack with any of the excluded nodes? If we had a couple of excluded nodes (eg decommissioning nodes) and then we pick some new used nodes already - that exlude list will container <decommissioning ones> + <already chosen>, and then it will attempt to pick a new node that does not share any of those racks, which is too restrictive, as we are happy to use the decommissioning racks again.

The Javadoc is not really clear to me:

```
   * @param ancestorGen If 0, then no same generation ancestor enforcement on
   *                     both excludedNodes and affinityNode. If greater than 0,
   *                     then apply to affinityNode(if not null), or apply to
   *                     excludedNodes if affinityNode is null
```

ancesorGen is getting passed as 1 so it is greater than zero.

Reply:
This java doc gives a bit more of a clear indication about what is going on I think:

```
  /**
   * Randomly choose a leaf node from <i>scope</i>.
   *
   * If scope starts with ~, choose one from the all nodes except for the
   * ones in <i>scope</i>; otherwise, choose nodes from <i>scope</i>.
   * If excludedNodes is given, choose a node that's not in excludedNodes.
   *
   * @param scope range of nodes from which a node will be chosen
   * @param excludedNodes nodes to be excluded from.
   * @param ancestorGen matters when excludeNodes is not null. It means the
   * ancestor generation that's not allowed to share between chosen node and the
   * excludedNodes. For example, if ancestorGen is 1, means chosen node
   * cannot share the same parent with excludeNodes. If value is 2, cannot
   * share the same grand parent, and so on. If ancestorGen is 0, then no
   * effect.
   *
   * @return the chosen node
   */
  Node chooseRandom(String scope, Collection<Node> excludedNodes,
      int ancestorGen);
```

It sounds like if you have affinityNodes passed and ancestorGen=1 (its always 1 in the code we are looking at) - then the selected node must share a rack with the affinityNodes.

However, if affinityNodes are null, and there are excluded nodes, then with ancestorGen=1 the selected nodes cannot share a rack with any nodes listed in excludedNodes.

I think that we will somehow have to pass the chosenNodes / usedNodes as excluded, and then use excludedScopes to exclude the "non-used" excluded nodes.

Reply:
I have updated to handle pass the chosenNodes / usedNodes as excluded, and then use excludedScopes to exclude the "non-used" excluded nodes.
@sodonnel , Can you please review.